x/ecocredit/server/basket/create.go

```diff
@@ -3,10 +3,120 @@ package basket
 import (
 	"context"
 
-	baskettypes "github.com/regen-network/regen-ledger/x/ecocredit/basket"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	basketv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1"
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/basket"
 )
 
-func (k Keeper) Create(ctx context.Context, create *baskettypes.MsgCreate) (*baskettypes.MsgCreateResponse, error) {
-	//TODO implement me
-	panic("implement me")
+func (k Keeper) Create(ctx context.Context, msg *basket.MsgCreate) (*basket.MsgCreateResponse, error) {
+	rgCtx := types.UnwrapSDKContext(ctx)
+	fee := k.ecocreditKeeper.GetCreateBasketFee(ctx)
+	if err := basket.ValidateCreateFee(msg, fee); err != nil {
+		return nil, err
+
+	}
+	sender, err := sdk.AccAddressFromBech32(msg.Curator)
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.distKeeper.FundCommunityPool(rgCtx.Context, fee, sender)
+	if err != nil {
+		return nil, err
+	}
+	if err = assertCreditTypeExists(ctx, k.ecocreditKeeper, msg.CreditTypeName, msg.Exponent); err != nil {
+		return nil, err
+	}
+
+	// TODO: need to decide about the denom creation
+	denom := msg.Name
+
+	id, err := k.stateStore.BasketStore().InsertReturningID(ctx, &basketv1.Basket{
+		BasketDenom:       denom,
+		DisableAutoRetire: msg.DisableAutoRetire,
+		CreditTypeName:    msg.CreditTypeName,
+		DateCriteria:      msg.DateCriteria.ToApi(),
+		Exponent:          msg.Exponent,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err = k.indexAllowedClasses(rgCtx, id, msg.AllowedClasses); err != nil {
+		return nil, err
+	}
+
+	k.bankKeeper.SetDenomMetaData(rgCtx.Context, banktypes.Metadata{
+		DenomUnits: []*banktypes.DenomUnit{
+			{
+				Denom:    msg.DisplayName,
+				Exponent: msg.Exponent,
+				Aliases:  nil,
+			},
+		},
+		Base:    denom,
+		Display: msg.DisplayName,
+		Name:    msg.DisplayName,
+		Symbol:  msg.DisplayName,
+	})
+
+	err = rgCtx.Context.EventManager().EmitTypedEvent(&basket.EventCreate{
+		BasketDenom: denom,
+		Curator:     msg.Curator,
+	})
+
+	return &basket.MsgCreateResponse{BasketDenom: denom}, err
+}
+
+// assertCreditTypeExists returns error if a given credit type doesn't exist or it's precision
+// is bigger then the requested exponent.
+func assertCreditTypeExists(ctx context.Context, k EcocreditKeeper, creditType string, exponent uint32) error {
```

Comment:
Maybe `validateCreditType` given we are checking more than whether or not the credit type exists.

Reply:
OK

Reply:
done